lib/xdr/assets/asset_code12.ex

```diff
@@ -4,46 +4,76 @@ defmodule Stellar.XDR.AssetCode12 do
   """
   @behaviour XDR.Declaration
 
-  @type t :: %__MODULE__{code: String.t()}
+  @type t :: %__MODULE__{code: String.t(), length: non_neg_integer()}
 
-  defstruct [:code]
+  defstruct [:code, :length]
 
-  @max_size 12
-
-  @opaque_spec XDR.VariableOpaque.new(nil, @max_size)
+  @length_range 5..12
 
   @spec new(code :: String.t()) :: t()
-  def new(code), do: %__MODULE__{code: code}
+  def new(code), do: %__MODULE__{code: code, length: byte_size(code)}
 
   @impl true
-  def encode_xdr(%__MODULE__{code: code}) do
+  def encode_xdr(%__MODULE__{code: code, length: length}) when length in @length_range do
     code
-    |> XDR.VariableOpaque.new(@max_size)
-    |> XDR.VariableOpaque.encode_xdr()
+    |> XDR.FixedOpaque.new(length)
+    |> XDR.FixedOpaque.encode_xdr()
   end
 
+  def encode_xdr(_asset_code), do: {:error, :invalid_length}
+
   @impl true
-  def encode_xdr!(%__MODULE__{code: code}) do
+  def encode_xdr!(%__MODULE__{code: code, length: length}) when length in @length_range do
     code
-    |> XDR.VariableOpaque.new(@max_size)
-    |> XDR.VariableOpaque.encode_xdr!()
+    |> XDR.FixedOpaque.new(length)
+    |> XDR.FixedOpaque.encode_xdr!()
   end
 
+  def encode_xdr!(_asset_code), do: raise(XDR.FixedOpaqueError, :invalid_length)
```

Comment:
Not sure if we should expose the `XDR.FixedOpaqueError`, which is an internal error returned by the XDR library. Something like `Stellar.XDR.AssetCode12Error` could be more appropriate. What do you think?

Reply:
@luishurtado definitely, agree with you!

Reply:
@luishurtado The Exceptions modules were implemented. Could you please check again? thanks!